### PR TITLE
HealthMonitor improvements (#2972)

### PR DESCRIPTION
**Pull request by mathewc**
Addresses https://github.com/Azure/azure-functions-host/issues/2972. With these changes, the behavor will be:

- if the host remains unhealthy for a period of time and host restarts aren't fixing it, an app shutdown is initiated (based on HostHealthMonitor config settings)
- before we shutdown, we write an app instance specific marker file to signal to the restarted app that a shutdown was initiated
- if after this shutdown the app remains unhealthy and a shutdown is triggered *again*, the presence of the marker file will cause us to initiate a *hard* shutdown

So in summary: previously we'd continue in a "soft" shutdown loop, recycling the app domain in an attempt to correct. Now, after 2 full cycles we'll escalate to a "hard" shutdown which will recycle the process.

I'll make the same changes in the v1 branch as well.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -66,6 +66,7 @@ public static void AddWebJobsScriptHost(this IServiceCollection services, IConfi
             services.AddStandbyServices();
 
             // Core script host services
+            services.TryAddSingleton<IScriptJobHostEnvironment, WebScriptJobHostEnvironment>();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs]: Had to move this from its previous location (see below). It turns out that we actually have a bug currently where the WebJobsScriptHostService environment is null on resolution.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 2]
```diff
@@ -59,6 +60,9 @@ public class WebJobsScriptHostService : IHostedService, IScriptHostManager, IDis
                 _healthCheckWindow = new SlidingWindow<bool>(_healthMonitorOptions.Value.HealthCheckWindow);
                 _hostHealthCheckTimer = new Timer(OnHostHealthCheckTimer, null, TimeSpan.Zero, _healthMonitorOptions.Value.HealthCheckInterval);
             }
+
+            // shutdown is scoped to a single host instance
+            _shutdownSentinelFilePath = Path.Combine(_applicationHostOptions.CurrentValue.LogPath, "Host", $"{Utility.GetInstanceId()}-{ScriptConstants.ShutdownSentinelFileName}");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Chose an instance based file approach here as the signal, rather than trying to play games with the AppDomainId/count. AppDomain count may go up for other reasons than restarts, e.g. language worker recycles. The instance prefix on this marker file means only the affected instance will recognize it.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 3]
```diff
@@ -341,18 +352,32 @@ internal bool IsHostHealthy(bool throwWhenUnhealthy = false)
...
             if (ShouldMonitorHostHealth && _healthCheckWindow.GetEvents().Where(isHealthy => !isHealthy).Count() > _healthMonitorOptions.Value.HealthCheckThreshold)
             {
-                // if the number of times the host has been unhealthy in
-                // the current time window exceeds the threshold, recover by
-                // initiating shutdown
-                var message = $"Host unhealthy count exceeds the threshold of {_healthMonitorOptions.Value.HealthCheckThreshold} for time window {_healthMonitorOptions.Value.HealthCheckWindow}. Initiating shutdown.";
-                _logger.LogError(0, message);
+                logger = logger ?? _logger;
                 var environment = _rootServiceProvider.GetService<IScriptJobHostEnvironment>();
-                environment.Shutdown();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: This environment was resolving to null before the service registration fix above

File: src/WebJobs.Script/ConsoleScriptJobHostEnvironment.cs [hunk 4]
```diff
@@ -20,12 +23,18 @@ public ConsoleScriptJobHostEnvironment(IApplicationLifetime applicationLifetime,
 
         public void RestartHost()
         {
-            _applicationLifetime.StopApplication();
+            if (_shutdownRequested == 0 && Interlocked.Exchange(ref _restartRequested, 1) == 0)
```
Comment (mathewc) [inline src/WebJobs.Script/ConsoleScriptJobHostEnvironment.cs]: Applying the same guard pattern here that we had for the other implementation
